src/unix/process.c

```diff
@@ -342,6 +345,239 @@ static void uv__process_child_init(const uv_process_options_t* options,
+                                           const uv_process_options_t* options,
+                                           int stdio_count,
+                                           int (*pipes)[2])
+{
+  int fd;
+  int err;
+
+  err = posix_spawn_file_actions_init(actions);
+  if (err != 0) {
+    /* If initialization fails, no need to de-init, just return */
+    return err;
+  }
+
+  /* Set the current working directory if requested */
+  if (options->cwd != NULL) {
+    err = posix_spawn_file_actions_addchdir_np(actions, options->cwd);
+    if (err != 0)
+      goto error;
+  }
+
+  /* First, dupe any required fd into orbit, out of the range of 
+   * the descriptors that should be mapped in. */
+  for(fd = 0 ; fd < stdio_count; ++fd) {
+    if (pipes[fd][1] < 0)
+      continue;
+    
+    err = posix_spawn_file_actions_adddup2(actions, pipes[fd][1], stdio_count + fd);
+    if (err != 0)
+      goto error;
+  }
+
+  /*  Second, move the descriptors into their respective places */
+  for(fd = 0 ; fd < stdio_count; ++fd) {
+    if (pipes[fd][1] < 0)
+      continue;
+
+    err = posix_spawn_file_actions_adddup2(actions, stdio_count + fd, fd);
+    if (err != 0)
+      goto error;
+  }
+
+  /*  Finally, close all the superfluous descriptors */
+  for(fd = 0; fd < stdio_count; ++fd) {
+    if (pipes[fd][1] < 0)
+      continue;
+    
+    err = posix_spawn_file_actions_addclose(actions, stdio_count + fd);
+    if (err != 0)
+      goto error;
+  }
+
+  /*  Finally process the standard streams as per de documentation */
+  for(fd = 0 ; fd < 3 ; ++fd) {
+    /*  If ignored, open as /dev/null */
+    const int oflags = fd == 0 ? O_RDONLY : O_RDWR;
+    const int mode = 0;
+
+    if (pipes[fd][1] != -1) 
+      continue;
+    
+    err = posix_spawn_file_actions_addopen(actions, fd, "/dev/null", oflags, mode);
+    if (err != 0)
+      goto error;
+  }  
+
+  return err;
+
+error:
+  (void) posix_spawn_file_actions_destroy(actions);
+  return err;
+}
+
+int uv__spawn_and_init_child_posix_spawn(const uv_process_options_t* options,
+                                         int stdio_count,
+                                         int (*pipes)[2], 
+                                         pid_t* pid) {
+  int err;
+  posix_spawnattr_t attrs;
+  posix_spawn_file_actions_t actions;
+
+  err = uv__spawn_set_posix_spawn_attrs(&attrs, options);
+  if (err != 0) 
+    goto error;
+
+  err = uv__spawn_set_posix_spawn_file_actions(&actions, options, stdio_count, pipes);
+  if (err != 0) {
+    (void) posix_spawnattr_destroy(&attrs);
+    goto error;
+  }
+
+  /*  Preserve parent environment if not explicitly set */
+  char** env = options->env ? options->env : environ;
+
+  /*  Spawn the child */
+  err = posix_spawnp(pid, options->file, &actions, &attrs, options->args, env);
+
+  /* Destroy the actions/attributes */
+  (void) posix_spawn_file_actions_destroy(&actions);
+  (void) posix_spawnattr_destroy(&attrs);
```

Comment:
Seemed to me that the failure of either of these functions was not a reason to signal failure to the caller.

Reply:
optional: I'd perhaps suggest making them call abort instead?

Reply:
Your call, @vtjnash. If you think that failure of these calls warrants termination, let's do that.